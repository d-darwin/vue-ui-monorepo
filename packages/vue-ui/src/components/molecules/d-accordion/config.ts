import dAccordionStyles from "./d-accordion.css";
import dDetailsStyles from "./d-details.css";

export default {
  name: "DAccordion",
  class: dAccordionStyles["dAccordion"],
  details: {
    ref: "detailsRef",
    name: "DDetails",
    class: dDetailsStyles["dDetails"],
  },
  summaryOptions: {
    class: dDetailsStyles["summary"],
  },
  summaryIcon: "â››",
  summaryAfterClass: dDetailsStyles["summaryAfter"],
  contentOptions: {
    key: "detailsContent",
    ref: "detailsContentRef",
    class: dDetailsStyles["content"],
  },
  provideInjectKey: "commonProps",
} as const;
