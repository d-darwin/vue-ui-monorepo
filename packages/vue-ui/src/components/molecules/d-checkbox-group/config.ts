export default {
  name: "DCheckboxGroup",
  className: "dCheckboxGroup",
  labelClassName: "label",
  errorClassName: "error",
} as const;
