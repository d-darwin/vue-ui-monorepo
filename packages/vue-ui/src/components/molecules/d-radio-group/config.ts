import { TYPE } from "@darwin-studio/vue-ui/src/components/atoms/d-caption/constant";
import styles from "./index.css?module";

export default {
  name: "DRadioGroup",
  class: styles["dRadioGroup"],
  labelClass: styles["label"],
  radioClass: styles["radio"],
  captionOptions: {
    type: TYPE.DANGER,
    class: styles["caption"],
  },
  captionOffset: "0.2em",
} as const;
