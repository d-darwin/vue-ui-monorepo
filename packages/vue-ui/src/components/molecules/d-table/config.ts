export default {
  name: "DTable",
  className: "dTable",
  rowClassName: "row",
} as const;
