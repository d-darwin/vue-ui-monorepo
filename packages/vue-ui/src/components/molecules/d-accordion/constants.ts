import { HTMLAttributes } from "vue";
import config from "./config";
import styles from "./d-details.css?module";

export const SUMMARY_DEFAULTS: HTMLAttributes & { key: string } = {
  key: "summary",
  class: styles[config.summaryClassName],
};

export const CONTENT_DEFAULTS: HTMLAttributes & { key: string; ref: string } = {
  key: "content",
  ref: config.contentRef,
  class: styles[config.contentClassName],
};

export const PROVIDE_INJECT_KEY = "commonProps";
