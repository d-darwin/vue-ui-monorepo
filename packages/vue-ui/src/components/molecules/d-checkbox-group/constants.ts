import type { DCaptionProps } from "@darwin-studio/vue-ui/src/components/atoms/d-caption/types";
import { TYPE as CAPTION_TYPE } from "@darwin-studio/vue-ui/src/components/atoms/d-caption/constant";
import config from "./config";
import styles from "./index.css?module";

export const CAPTION_DEFAULTS: DCaptionProps = {
  type: CAPTION_TYPE.DANGER,
  class: styles[config.captionClassName],
};
