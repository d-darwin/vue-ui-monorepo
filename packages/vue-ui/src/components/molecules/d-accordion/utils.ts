import { Transition } from "@darwin-studio/ui-codegen/dist/types/transition";
import { TRANSITION_VALUE } from "@darwin-studio/ui-codegen/dist/constants/transition";
import getConstantKey from "@darwin-studio/vue-ui/src/utils/get-constant-key";

export function getTransitionDuration(transition?: Transition): number {
  if (!transition) return 0;

  const transitionKey = getConstantKey(
    transition
  ) as keyof typeof TRANSITION_VALUE;

  return TRANSITION_VALUE[transitionKey]?.duration || 0;
}
