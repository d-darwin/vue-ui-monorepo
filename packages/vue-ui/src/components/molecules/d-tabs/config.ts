import { COLOR_SCHEME } from "@darwin-studio/ui-codegen/dist/constants/color-scheme";
import { SIZE } from "@darwin-studio/ui-codegen/dist/constants/size";
import { TAG_NAME } from "@darwin-studio/vue-ui/src/constants/tag-name";
import dTabsStyles from "./d-tabs.css?module";
import dTabStyles from "./d-tab.css?module";
import dTabpanelStyles from "./d-tabpanel.css?module";

export default {
  tabsName: "DTabs",
  tabsClass: dTabsStyles.dTabs,
  provideInjectKey: "tabsProvided",
  injectedActiveIdPath: "injection.activeId",
  colorScheme: COLOR_SCHEME.PRIMARY,
  size: SIZE.MEDIUM,

  tablistOptions: {
    role: "tablist",
    class: dTabsStyles.dTablist,
  },
  tablistTag: TAG_NAME.UL,
  tablistRef: "tablist",

  tabName: "DTab",
  tabOptions: {
    role: "tab",
    class: dTabStyles.dTab,
  },
  tabTag: TAG_NAME.LI,

  tabpanelName: "DTabpanel",
  tabpanelOptions: {
    role: "tabpanel",
    tabindex: 0,
    class: dTabpanelStyles.dTabpanel,
  },
  tabpanelTag: TAG_NAME.DIV,
};
