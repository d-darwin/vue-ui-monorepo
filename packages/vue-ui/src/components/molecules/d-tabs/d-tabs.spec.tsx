import { mount } from "@vue/test-utils";
import { baseClassCase, tagCase } from "@/utils/test-case-factories";
import { FONT } from "@darwin-studio/ui-codegen/dist/constants/font";
import { SIZE } from "@darwin-studio/ui-codegen/dist/constants/size";
import { PADDING } from "@darwin-studio/ui-codegen/dist/constants/padding";
import { TRANSITION } from "@darwin-studio/ui-codegen/dist/constants/transition";
import { DTabs, DTab, DTabpanel } from "./index";
import config from "./config";

jest.mock("chalk", () => ({
  greenBright: jest.fn(),
  red: jest.fn(),
  yellowBright: jest.fn(),
}));

describe("DTabs", () => {
  const wrapper = mount(DTabs);

  baseClassCase(wrapper, config.tabsClass);

  it("Should render config.tablistOptions.className to the tablist class name", () => {
    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    expect(tablist).toBeTruthy();
  });

  it("Should render props.tablistLabel to the tablist aria-label attr", async () => {
    const tablistLabel = "Some aria label";
    await wrapper.setProps({
      tablistOptions: { ["aria-label"]: tablistLabel },
    });
    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    expect(tablist.attributes("aria-label")).toBe(tablistLabel);
  });

  it("Should props.tablistOptions.class to the tablist class name", async () => {
    const tablistClass = "some-tablist-class";
    await wrapper.setProps({ tablistOptions: { class: tablistClass } });
    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    expect(tablist.classes()).toContain(tablistClass);
  });

  it("Should render props.tabs as content of the tablist", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      props: {
        tabs: [
          <DTab id="111" tabpanelId="111-1" active={true} content={"Tab 1"} />,
          <DTab id="222" tabpanelId="222-2" content={"Tab 2"} />,
        ],
        tabpanels: [
          <DTabpanel tabId="111" active={true} content={"Tabpanel 1"} />,
          <DTabpanel tabId="222" content={"Tabpanel 2"} />,
        ],
      },
    });

    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    const tabs = tablist.findAllComponents(DTab);
    expect(tabs?.[0]?.element).toMatchSnapshot(); // TODO: find out other way to compare
    expect(tabs?.[1]?.element).toMatchSnapshot(); // TODO: find out other way to compare
  });

  it("Should render slots.tabs as content of the tablist", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      slots: {
        tabs: [
          <DTab id="111" tabpanelId="111-1" active={true} content={"Tab 1"} />,
          <DTab id="222" tabpanelId="222-2" content={"Tab 2"} />,
        ],
        tabpanels: [
          <DTabpanel
            tabId="111"
            id="111-1"
            active={true}
            content={"Tabpanel 1"}
          />,
          <DTabpanel tabId="222" id="222-2" content={"Tabpanel 2"} />,
        ],
      },
    });

    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    const tabs = tablist.findAllComponents(DTab);
    expect(tabs?.[0]?.element).toMatchSnapshot(); // TODO: find out other way to compare
    expect(tabs?.[1]?.element).toMatchSnapshot(); // TODO: find out other way to compare
  });

  it("Should inject props.size to commonProps.size of containing DTab components", async () => {
    const size = SIZE.LARGE;
    await wrapper.setProps({
      size,
      tabs: [<DTab id="111" active={true} content={"Tab 1"} />],
    });
    const tabWrapper = wrapper.findComponent(DTab);
    expect(tabWrapper.vm.commonProps?.size).toBe(size);
  });

  it("Should render props.tabpanels as content of the list of tabpanels", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      props: {
        tabs: [
          <DTab id="111" active={true} content={"Tab 1"} />,
          <DTab id="222" tabpanelId="222-2" content={"Tab 2"} />,
        ],
        tabpanels: [
          <DTabpanel tabId="111" active={true} content={"Tabpanel 1"} />,
          <DTabpanel tabId="222" content={"Tabpanel 2"} />,
        ],
      },
    });

    const tabpanels = wrapper.findAllComponents(DTabpanel);
    expect(tabpanels?.[0]?.element).toMatchSnapshot(); // TODO: find out other way to compare
    expect(tabpanels?.[1]?.element).toMatchSnapshot(); // TODO: find out other way to compare
  });

  it("Should render slots.tabpanels as content of the list of tabpanels", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      slots: {
        tabs: [
          <DTab id="111" tabpanelId="111-1" active={true} content={"Tab 1"} />,
          <DTab id="222" tabpanelId="222-2" content={"Tab 2"} />,
        ],
        tabpanels: [
          <DTabpanel
            tabId="111"
            id="111-1"
            active={true}
            content={"Tabpanel 1"}
          />,
          <DTabpanel tabId="222" id="222-2" content={"Tabpanel 2"} />,
        ],
      },
    });

    const tabpanels = wrapper.findAllComponents(DTabpanel);
    expect(tabpanels?.[0]?.element).toMatchSnapshot(); // TODO: find out other way to compare
    expect(tabpanels?.[1]?.element).toMatchSnapshot(); // TODO: find out other way to compare
  });

  it("Should inject props.size to commonProps.size of the DTabpanel components", async () => {
    const size = FONT.SMALL;
    await wrapper.setProps({
      size,
      tabpanels: [<DTabpanel id="222" tabId="222-2" content={"Tabpanel 2"} />],
    });
    const tabpanelWrapper = wrapper.findComponent(DTabpanel);
    expect(tabpanelWrapper.vm.commonProps?.size).toBe(size);
  });

  it("Should pass props.disabled to the containing DTabs", async () => {
    const disabled = true;
    await wrapper.setProps({
      disabled,
      tabs: [
        <DTab id="111" active={true} content={"Tab 1"} />,
        <DTab id="222" disabled={false} content={"Tab 2"} />,
      ],
    });
    const tabs = wrapper.findAllComponents(DTab);
    expect(tabs[0]?.vm.commonProps.disabled).toBe(disabled);
    expect(tabs[1]?.vm.commonProps.disabled).toBe(true);
  });

  // TODO paddingClassesCase(wrapper, wrapper.findComponent(DTab));
  it(`Should render props.padding to DTab padding classes`, async () => {
    const padding = PADDING.EQUAL;
    const wrapper = mount(DTabs, {
      props: {
        padding,
        tabs: [<DTab id="not empty" content={"Tab 1"} />],
      },
    });
    const tab = wrapper.findComponent(DTab);
    expect(tab.vm.commonProps.padding).toBe(padding);
  });

  // TODO: transitionClassCase(wrapper, wrapper.findComponent(DTab))
  it(`Should render props.transition to DTab transition class`, async () => {
    const transition = TRANSITION.AVERAGE;
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      props: {
        transition,
        tabs: [<DTab id={4345345345345} content={"Tab 1"} />],
      },
    });
    const tab = wrapper.findComponent(DTab);
    expect(tab.vm.commonProps.transition).toBe(transition);
  });

  tagCase(wrapper);

  it("Should render tablist as element passed in props.tablistTag", async () => {
    const tablistTag = "section";
    await wrapper.setProps({ tablistTag });
    const tablist = wrapper.find(`.${config.tablistOptions.class}`);
    expect(tablist.element.tagName).toEqual(tablistTag.toLocaleUpperCase());
  });

  it("Should focus on prev tab on ArrowLeft keydown", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      props: {
        tabs: [
          <DTab id={12} content={"Tab 1"} />,
          <DTab id={23} content={"Tab 2"} />,
        ],
      },
    });

    const tabs = wrapper.findAllComponents(DTab);
    const tab0 = tabs?.[0];
    tab0.vm.$el.focus = jest.fn();
    const tab1 = tabs?.[1];
    tab1.vm.$el.focus = jest.fn();

    await tab0?.trigger("focus");
    await tab0?.trigger("keydown", { key: "ArrowLeft" });
    expect(tab1.vm.$el.focus).toBeCalled();

    await tab1?.trigger("focus");
    await tab1?.trigger("keydown", { key: "ArrowLeft" });
    expect(tab0.vm.$el.focus).toBeCalled();
  });

  it("Should focus on next tab on ArrowRight keydown", async () => {
    // TODO: reactivity is broken ???
    const wrapper = mount(DTabs, {
      props: {
        tabs: [
          <DTab id={1} content={"Tab 1"} />,
          <DTab id={2} content={"Tab 2"} />,
        ],
      },
    });

    const tabs = wrapper.findAllComponents(DTab);
    const tab0 = tabs?.[0];
    tab0.vm.$el.focus = jest.fn();
    const tab1 = tabs?.[1];
    tab1.vm.$el.focus = jest.fn();

    await tab0?.trigger("focus");
    await tab0?.trigger("keydown", { key: "ArrowRight" });
    expect(tab1.vm.$el.focus).toBeCalled();

    await tab1?.trigger("focus");
    await tab1?.trigger("keydown", { key: "ArrowRight" });
    expect(tab0.vm.$el.focus).toBeCalled();
  });

  it("Should call props.whenChange with focused tab id on Enter", async () => {
    // TODO: reactivity is broken ???
    const whenChange = jest.fn();
    const tabId0 = "a";
    const tabId1 = "b";
    const wrapper = mount(DTabs, {
      props: {
        tabs: [
          <DTab id={tabId0} content={"Tab 1"} />,
          <DTab id={tabId1} content={"Tab 2"} />,
        ],
        whenChange,
      },
    });

    const tabs = wrapper.findAllComponents(DTab);
    const tab0 = tabs?.[0];
    const tab1 = tabs?.[1];

    await tab0?.trigger("focus");
    await tab0?.trigger("keydown", { key: "Enter" });
    expect(whenChange).toBeCalledWith(tabId0);

    await tab1?.trigger("focus");
    await tab1?.trigger("keydown", { key: "Enter" });
    expect(whenChange).toBeCalledWith(tabId1);
  });

  it("Shouldn't do anything on keydown if there are less than 2 DTabs", async () => {
    const whenChange = jest.fn();
    const wrapper = mount(DTabs, {
      props: {
        tabs: [<DTab id={12} content={"Tab 1"} />],
        whenChange,
      },
    });

    const tab = wrapper.findComponent(DTab);
    await tab?.trigger("focus");
    await tab?.trigger("keydown", { key: "Enter" });
    expect(whenChange).not.toHaveBeenCalled();
  });

  it("Shouldn't do anything on keydown if DTabs haven't id", async () => {
    const whenChange = jest.fn();
    const wrapper = mount(DTabs, {
      props: {
        tabs: [<DTab content={"Tab 1"} />, <DTab content={"Tab 2"} />],
        whenChange,
      },
    });

    const tab = wrapper.findComponent(DTab);
    await tab?.trigger("focus");
    await tab?.trigger("keydown", { key: "Enter" });
    await tab?.trigger("keydown", { key: "ArrowLeft" });
    await tab?.trigger("keydown", { key: "ArrowRight" });
    expect(whenChange).not.toHaveBeenCalled();
  });

  // TODO: Should following to be a part of the DTabs component:
  // DTabpanel.active ???
  // DTab.active ???
  // DTab.onClick ???
  // DTab.whenClick ???
});
