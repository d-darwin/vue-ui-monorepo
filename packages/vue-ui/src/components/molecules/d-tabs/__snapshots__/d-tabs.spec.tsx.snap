// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DTabs Should render props.tabpanels as content of the list of tabpanels 1`] = `
<div
  aria-expanded="true"
  class="dTabpanel fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium transitionFast"
  role="tabpanel"
  tabindex="0"
>
  Tabpanel 1
</div>
`;

exports[`DTabs Should render props.tabpanels as content of the list of tabpanels 2`] = `
<div
  aria-hidden="true"
  class="dTabpanel fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium transitionFast"
  hidden=""
  role="tabpanel"
  tabindex="0"
>
  Tabpanel 2
</div>
`;

exports[`DTabs Should render props.tabs as content of the tablist 1`] = `
<li
  aria-controls="111-1"
  aria-selected="true"
  class="dTab fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium sizeMedium transitionFast active"
  id="111"
  role="tab"
  tabindex="0"
>
  Tab 1
</li>
`;

exports[`DTabs Should render props.tabs as content of the tablist 2`] = `
<li
  aria-controls="222-2"
  class="dTab fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium sizeMedium transitionFast"
  id="222"
  role="tab"
  tabindex="-1"
>
  Tab 2
</li>
`;

exports[`DTabs Should render slots.tabpanels as content of the list of tabpanels 1`] = `
<div
  aria-expanded="true"
  class="dTabpanel fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium transitionFast"
  role="tabpanel"
  tabindex="0"
>
  Tabpanel 1
</div>
`;

exports[`DTabs Should render slots.tabpanels as content of the list of tabpanels 2`] = `
<div
  aria-hidden="true"
  class="dTabpanel fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium transitionFast"
  hidden=""
  role="tabpanel"
  tabindex="0"
>
  Tabpanel 2
</div>
`;

exports[`DTabs Should render slots.tabs as content of the tablist 1`] = `
<li
  aria-controls="111-1"
  aria-selected="true"
  class="dTab fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium sizeMedium transitionFast active"
  id="111"
  role="tab"
  tabindex="0"
>
  Tab 1
</li>
`;

exports[`DTabs Should render slots.tabs as content of the tablist 2`] = `
<li
  aria-controls="222-2"
  class="dTab fontMedium outlinePrimaryMedium paddingDefault paddingDefaultMedium sizeMedium transitionFast"
  id="222"
  role="tab"
  tabindex="-1"
>
  Tab 2
</li>
`;
